src/main/java/org/apache/sling/api/resource/FallbackValueMap.java

```diff
@@ -0,0 +1,216 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.sling.api.resource;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.apache.sling.api.wrappers.ValueMapDecorator;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public class FallbackValueMap implements ValueMap {
+    List<Resource> resources;
+
+    Set<Entry<String, Object>> entrySet;
+    Set<String> keySet;
+    Set<String> accessed = new HashSet<>();
+    boolean cacheFilled = false;
+    Collection<Object> values;
+    ValueMap cache = new ValueMapDecorator(new HashMap<>());
+
+    private static final String EMPTY = "";
+
+    private static final String IMMUTABLE_ERROR_MESSAGE = "Fallback value maps are immutable";
+
+    /**
+     * Main entry point for building the value map's list.
+     *
+     * @param resources var args resources from where value map will be computed, order matters
+     *                  here : will be the order of the lookup.
+     */
+    public void setResources(Resource... resources) {
```

Comment:
I would make FallbackValueMap immutable, i.e. only allow setting the resources via a constructor. In fact, I think we should supply ValueMap instances to a constructor. For convenience there could be a second constructor taking Resource instances or a static factory method that gets each resource's ValueMap and passes them to the constructor (I prefer the latter).